core/src/server/rpc_module.rs

```diff
@@ -569,14 +569,15 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
 	}
 
 	/// Register a new asynchronous RPC method, which computes the response with the given callback.
-	pub fn register_async_method<R, Fun, Fut>(
+	pub fn register_async_method<R, E, Fun, Fut>(
```

Comment:
I am wondering whether we'd gain a bit more flexibility ultimately by copying Axum a bit and having a trait which is something like

```rust
trait IntoResponse {
    fn into_response(self) -> MethodResponse;
}
```

This could then be impl'd for `Result<V, E> where V: Serialize, E: Into<Error>` for instance:

```rust
impl <V, E> IntoResponse for Result<V, E> 
where
    V: Serialize,
    E: Into<Error>
{
    fn into_response(self) -> MethodResponse {
        // look at impl MethodResponse::response/error and use this here
    }
} 
```

fns like this `register_method` would expect a callback which was like `Fn(Params, &Context) ->  R` (where `R: IntoResponse`) and would call `into_response()` on the result (perhaps passing in any specific bits they need to as args to that).

Then we could impl things like `From<Infallible> for Error` to allow `Result<V, Infallible>` responses for instance, and perhaps have more room then to consider supporting other response types than Result in the future.

Just thinking out loud really..

Reply:
Oh yes, I considered it, but ultimately decided against doing huge changes in a single PR. This is something that, I think, can be added later iteratively - and, also, maybe someone else would like to tackle that instead of me. :D I'm fine with looking into it, but I'm also happy to share the fun.

Reply:
I propose that we split the error layers into jsonrpsee's and application level, and, further, split the jsonrpsee's error from one single struct into separate `ClientError` and `ServerError` as there are very distinctly different in handling. I think it would significantly improve the ergonomics, and unlock more explicit and sensible error customization for the client side, as this PR only addresses the server side really.

Reply:
If you like the direction I proposed I'll create an issue for splitting the errors and mark this discussion as resolved.

Reply:
@niklasad1 whatcha reckon? I think you were less keen on splitting the errors?

But in any case, I think for now I'm fine with just allowing the custom error and not being more general; perhaps in some future iteration we can add the above sort of thing if it makes sense :)

Reply:
I think it will be weird to keep the proc macro API i.e, to share the same definition with different error types for the client and server. It's still possible to do it in the proc macro code as you have changed it this PR.

Example both client and server
```rust
struct CustomError;

#[rpc(server, client)]
pub trait Rpc
{
	// The client will return `ClientError` here.....
	async fn f(&self) -> Result<String, CustomError>;
}
```

Example both cliet
```rust

#[rpc(client)]
pub trait Rpc
{
	// I can enter whatever type I want here but it will return ClientError anyway.....
	async fn f(&self) -> Result<String, ClientError>;
}
```

I mostly care about having the client and server to share the API definition which I like to avoid a bunch of boiler plate and so on.

Reply:
I'm thinking about it more like this:

```
struct CustomError;

#[rpc(server, client)]
pub trait Rpc {
	// The client will return `ClientError` here.....
	async fn f(&self) -> Result<String, CustomError>;
}


struct Server;

#[async_trait]
impl RpcServer for Server {
    async fn f(&self) -> Result<String, ServerError<CustomError>>; // <------ see the `ServerError`
}
```

And the client would have a `ClientError` or `ClientError<CustomError>` type.

Reply:
Sure, that idea is interesting but it complicates things in other crates such as `core` but I would prefer to keep it out of this PR.

That alone would cause plenty of changes

The benefit of splitting it up is that is easier to reason about the error scenarios but forces users to import a few different error types (fine though)